### ISSUE TITLE
Update subscription if subscription already exists

### ISSUE THREAD

**Issue by kzangeli** [telefonicaid/fiware-orion#1117]
Many (at least one :-)) users have applications that save their own state. E.g. they depend on a subscription or two in the broker. When an application like this starts, the first thing they do is to create the "subscription or two" they need.
Problem is, the broker doean't die only becuase their application dies (and even if it does, the subscriptions are persistent in mongo), so when that application dies and starts again, those "subscription or two", are sent to the broker again ...
Those subscriptions already existed but as that application is a little stateless, the peticion to create the subscriptions are sent to the broker again.
This is probably VERY common. We cannot assume that all applications using the broker will maintain some kind of state in some database. No, the broker needs to be a little more intelligent and not recreate the same subscription for the same recevier (reference, if we speak NGSI).

Not too hard to implement, but also far from trivial - what is the criteria to decide that a subscription already exists (100% match would suffice) ... perhaps we should go with 100% ...

**Comment by fgalan** [comment]
> Not too hard to implement, but also far from trivial - what is the criteria to decide that a subscription already exists (100% match would suffice) ... perhaps we should go with 100%

First version could be 100% matching. Next versions could try to improve the "same subscription" heuristic.

**Comment by fgalan** [comment]
Comment by @jmcanterafonseca in (duplicated) #1122:

> We need to specify the behavior so that if subscription refers to the same subject (entity set) and end point, a new subscription is not created but the current one updated. In that case the HTTP Response must be different than 201 ... need to check what will be and to specify that in the apib document.

**Comment by drypatrick** [comment]
Is the issue still actual? It's there any way to prevent duplication?

**Comment by fgalan** [comment]
I think the issue is sill actual. However, it is not a very important one at the present time. Client can prevent duplication analysing the existing subscriptions before creating a new one.

**Comment by drypatrick** [comment]
Yes but since the `id` of the subscription is auto-generated is not that simple to check the existence and then delete them.

**Comment by fgalan** [comment]
Yes, you have to inspects other fields (`subject`, etc.), which is easy (it is just to parse a JSON in a way or another).

Note that duplications semantics can depend on the client in its particular use case (for instance, should two subscriptions which differ only in `q` filter be considered equal? For some clients the answer could be "yes" and for others "no"). That's make my think that is the client and not Orion the one that should check this.

**Comment by tzzed** [comment]
Hi @fgalan 

You're right, using `?options=rejectIfExists` makes much more sense than adding a field to the payload.

That said, comparing full subscription JSON objects to detect duplicates can be expensive and error-prone (e.g., due to field ordering or irrelevant differences). To optimize this, maybe we could allow a more explicit comparison strategy, like:

```sh
POST /v2/subscriptions?options=checkIfExist&entities=Room&url=http://notify
```
This would let the broker check for existing subscriptions based on a minimal, well-defined set of fields (like target entities and notification URL), instead of comparing the entire subscription body. Just an idea to consider for better performance and control.
